### ISSUE TITLE
Mockery::on() pass-by-reference behavior not preserved

### ISSUE THREAD

**Issue by ahoffmann-wi** [php-mock/php-mock-mockery#2]
## Code Sample

*Some code has been omitted. Please let me know if you need clarification.*

```php
namespace edu\wisc\doit\lists\wisclist;

class WiscList_DB
{
    public function execute($sth, $select = true)
    {
        $r = oci_execute($sth);
        if ($r === false) {
            $e = oci_error($sth);
            throw new DbQueryException($e[static::MESSAGE]);
        }
        $output = array();
        if ($select === true) {
            oci_fetch_all($sth, $output);
        }
        return $output;
    }
}

class WiscList_DBTest extends \PHPUnit_Framework_TestCase
{
    public function testExecuteReturnsDataWhenSelectTrue()
    {
        PHPMockery::mock(__NAMESPACE__, 'oci_execute')->with(anything())->andReturn(true)->once();
        PHPMockery::mock(__NAMESPACE__, 'oci_fetch_all')->with(
            anything(),
            \Mockery::on(function (array &$output) {
                $output = array('ID' => array(1));
                return true;
            })
        )->once();
        $data = static::getWiscListDbWithMockConfig()->execute(null, true);
        $this->assertEquals(array('ID' => array(1)), $data);
    }

    private static function getWiscListDbWithMockConfig()
    {
        return new WiscList_DB(static::getMockConfig());
    }
}
```

## PHPUnit Test Output

```text
PHPUnit 4.8.13 by Sebastian Bergmann and contributors.

Runtime:	PHP 5.6.15 with Xdebug 2.3.3
Configuration:	/Users/ahoffmann/IdeaProjects/wisclist-admin/phpunit.xml

F

Time: 119 ms, Memory: 7.75Mb

There was 1 failure:

1) edu\wisc\doit\lists\wisclist\WiscList_DBTest::testExecuteReturnsDataWhenSelectTrue
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'ID' => Array (...)
 )

/Users/ahoffmann/IdeaProjects/wisclist-admin/tests/WiscList_DBTest.php:168

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.
```

## Explanation

The goal is to mock the [oci_fetch_all](http://php.net/manual/en/function.oci-fetch-all.php) method, which takes an array **by reference** and fills it with the data that was fetched. I am using `Mockery::on()` to [preserve pass-by-reference behavior](http://docs.mockery.io/en/latest/reference/pass_by_reference_behaviours.html). I have successfully done this in the past with vanilla Mockery. However, when using php-mock's integration with Mockery, the behavior is not preserved.

## Environment

* PHPUnit 4.8
* PHP 5.5
* Mockery 0.9.4
* php-mock 1.0.0
* php-mock-mockery 1.0.0

## Troubleshooting

* Confirmed that "vanilla" Mockery::on() preserves pass-by-reference behavior
* Ran a debugger to watch the arguments being passed. The `$output` variable is modified in Mockery::on(), but is reset back to the original argument sometime after the Mockery::on() method returns.
* Tried PHP 5.6 and PHPUnit 5, both still failed.

**Comment by malkusch** [comment]
Thank you very much for putting your time into this perfect report. I'll have a look into it. Meanwhile if you can't wait let me tell you that pass-by-reference is possible with `andReturnUsing(callable)`. That's already covered by a test case.

**Comment by malkusch** [comment]
In e40b4702c5f4d8c33de596ddfd2ddb4fc5376f5d I added a test to confirm your issue. Surprisingly the test [passed](https://travis-ci.org/php-mock/php-mock-mockery/builds/90410758). So your issue seems to be different.

My next guess is that you hit [Bug #68541](https://bugs.php.net/bug.php?id=68541). Could you confirm that by calling `\Mockery::close()` before `assertEquals()` to see if the actual mock was used? A cure against that is using [`Mock::define()`](http://php-mock.github.io/php-mock/api/class-phpmock.Mock.html#_define). Unfortunately I didn't put that into the `PHPMockery` api yet (I recently received a very similar report in https://github.com/php-mock/php-mock-phpunit/issues/4). So until I didn't extend `PHPMockery` you'll have to put something like this into your test:
```
/**
 * @beforeClass
 */
public static function definePHPMocks()
{
    $oci_fetch_all = new \phpmock\Mock(__NAMESPACE__, "oci_fetch_all", function () {});
    $oci_fetch_all->define();
}
```

**Comment by malkusch** [comment]
[`php-mock-mockery-1.1.0`](https://github.com/php-mock/php-mock-mockery/releases/tag/1.1.0) now has [`PHPMockery::define()`](http://php-mock.github.io/php-mock-mockery/api/class-phpmock.mockery.PHPMockery.html#_define):
```php
/**
 * @beforeClass
 */
public static function definePHPMocks()
{
    PHPMockery::define(__NAMESPACE__, "oci_fetch_all")
}
```

**Comment by ahoffmann-wi** [comment]
I will be testing your workaround this weekend (sorry for delay).

**Comment by ahoffmann-wi** [comment]
My test is working now with no changes! I was still using php-mock-mockery v1.0.0 when the test started working. Is the change you made in one of its dependencies? My test also worked with v1.1.0

Thanks for your help with this! And I again apologize for the delay in getting back to you.

**Comment by ahoffmann-wi** [comment]
Just a followup. For some reason, my test above worked without needing the `PHPMockery::define()` fix in PHP 5.5. However, in PHP 5.6, the test broke with the same error, but putting in `PHPMockery::define()` fixed it. Weird!  

Maybe there is something else different with the two environments I used. Also, I know PHP 5.6 introduced some new features to the `namespace` keyword. Regardless, this fix works.
